Parameterize "node" and "namespace" label in Prom queries

thomasvn opened the pull request:
## What does this PR change?
* OpenCost currently defaults to keying on the "node" and "namespace" label when querying Prom metrics. This PR adds additional configurability for this label name when querying Prom.
* Configurable via the `PROM_NODE_LABEL` and `PROM_NAMESPACE_LABEL` environment variables
* Note: These changes were built on directly top of the `v1.108` branch for immediate testing. Eventually this PR should be closed, and a new PR should be written which is built on top of the `develop` branch.

## Does this PR relate to any other PRs?
* Extends https://github.com/opencost/opencost/pull/2435.

## How will this PR impact users?
* By default, there is no impact.
* If users perform metric relabeling on any OpenCost, KubeStateMetrics, CAdvisor, or NodeExporter metrics, then this provides configurability for those labels.

## Does this PR address any GitHub or Zendesk issues?
* None

## How was this PR tested?
* TODO

## Does this PR require changes to documentation?
* TODO

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next OpenCost release? If not, why not?
* TODO

thomasvn committed:
Paramaterize 'node' label in Prom queries. Completed for Assets, but still need to work on 'allocation.go' and 'costmodel.go'

Signed-off-by: thomasvn <thomasnguyen96@gmail.com>

thomasvn committed:
Parameterize "node" label in prom queries in costmodel.go

Signed-off-by: thomasvn <thomasnguyen96@gmail.com>

thomasvn committed:
Parameterized "node" label in format strings for allocation.go. Also
identified all locations where we key on "node" and replaced that with
the new env var.

Signed-off-by: thomasvn <thomasnguyen96@gmail.com>

thomasvn committed:
Parameterize for resultNodeKey()

Signed-off-by: thomasvn <thomasnguyen96@gmail.com>

thomasvn committed:
Paramterize "namespace" label in allocation.go and cluster.go

Signed-off-by: thomasvn <thomasnguyen96@gmail.com>

thomasvn committed:
Parameterize "node" label in all locations that read Prom query results.

Signed-off-by: thomasvn <thomasnguyen96@gmail.com>

AjayTripathy commented on pkg/costmodel/costmodel.go:
```diff
@@ -144,50 +144,50 @@ const (
-					sum_over_time(kube_pod_container_resource_requests{resource="cpu", unit="core", container!="",container!="POD", node!="", %s}[%s] %s)
-				) by (namespace,container,pod,node,%s) , "container_name","$1","container","(.+)"
+					sum_over_time(kube_pod_container_resource_requests{resource="cpu", unit="core", container!="",container!="POD", %s!="", %s}[%s] %s)
+				) by (namespace,container,pod,%s,%s) , "container_name","$1","container","(.+)"
```
parameterize namespace here

AjayTripathy commented on pkg/costmodel/costmodel.go:
```diff
@@ -144,50 +144,50 @@ const (
-					sum_over_time(kube_pod_container_resource_requests{resource="memory", unit="byte", container!="",container!="POD", node!="", %s}[%s] %s)
-				) by (namespace,container,pod,node,%s) , "container_name","$1","container","(.+)"
+					sum_over_time(kube_pod_container_resource_requests{resource="memory", unit="byte", container!="",container!="POD", %s!="", %s}[%s] %s)
+				) by (namespace,container,pod,%s,%s) , "container_name","$1","container","(.+)"
```
parameterize namespace

AjayTripathy commented on pkg/costmodel/costmodel.go:
```diff
@@ -214,8 +214,8 @@ const (
-				sum_over_time(container_cpu_allocation{container!="",container!="POD", node!="", %s}[%s])
-			) by (namespace,container,pod,node,%s) * %f / 60 / 60
+				sum_over_time(container_cpu_allocation{container!="",container!="POD", %s!="", %s}[%s])
+			) by (namespace,container,pod,%s,%s) * %f / 60 / 60
```
parameterize namespace

thomasvn committed:
Check the default label name if the custom/relabeled label is not found.

Signed-off-by: thomasvn <thomasnguyen96@gmail.com>

thomasvn committed:
Paramaterize 'namespace' label in all Prom queries in costmodel.go,
except for cAdvisor and GPU queries.

Signed-off-by: thomasvn <thomasnguyen96@gmail.com>

thomasvn committed:
Temporarily set default env to 'exported_node' and 'exported_namespace'

Signed-off-by: thomasvn <thomasnguyen96@gmail.com>

thomasvn committed:
Revert changes, since they cause unit tests to fail.

Signed-off-by: thomasvn <thomasnguyen96@gmail.com>

thomasvn committed:
New func GetDefaultPromNamespaceLabel() which applies to metrics which
will always have the default "namespace" metric label according to
scrapeconfig.

Signed-off-by: thomasvn <thomasnguyen96@gmail.com>

thomasvn committed:
Missed paramaterizing 'node' in one location. Switched one metric label
from default namespace to env namespace.

Signed-off-by: thomasvn <thomasnguyen96@gmail.com>

thomasvn committed:
Fix namespace label for pvcMap

Signed-off-by: thomasvn <thomasnguyen96@gmail.com>

thomasvn committed:
Prevent overwriting of node labels in the map.

Signed-off-by: thomasvn <thomasnguyen96@gmail.com>

cklingspor commented:
Hi @thomasvn ,
I was reading through this  and https://github.com/opencost/opencost/pull/2423 since Im running into this issue and also came up with the solution from #2423. Why was this closed?

thomasvn commented:
HI @cklingspor! This PR's scope was much larger than the issue described in #2423. It eventually became too environment specific and needed extra caution before being merged. I went ahead and closed this PR for now, but I agree that we should revisit #2423 and find a solution there.
